### PR TITLE
Rest Consumer Support

### PR DESCRIPTION
**Pull request by dartartem**
### Added support for rest consumer.

**Rest consumer is declared using properties:**

```
subscription.message.s1.subscriberId=SomeSubscriber
subscription.message.s1.channels=test-channel
subscription.message.s1.callbackSubscriptionId=SomeSubscription
eventuate.rest.consumer.base.url=http://${DOCKER_HOST_IP:localhost}:8081/messages
```

- It subscribes to `test-channel` using `SomeSubscriber` as subscriber id.
- `s1` is name of subscription, used only for grouping can be any.
- `SomeSubscription` is custom subscription id, used for identifies message handler.

**Declaring controller:**

```
@RestController
public class TestController {
  @PostMapping(path = "/messages/SomeSubscription")
  public void handleMessages(@RequestBody HttpMessage httpMessage) {
    ...
  }
}
```

Initial implementation of proxy service generated unique id for any subscription:

https://github.com/eventuate-tram/eventuate-tram-messaging-http/blob/master/eventuate-tram-messaging-proxy-service/src/main/java/io/eventuate/tram/messaging/proxy/service/SubscriptionService.java#L46-L64

And used for identifying internal http handlers used for http consumer:

https://github.com/eventuate-tram/eventuate-tram-messaging-http/blob/master/eventuate-tram-consumer-http-micronaut/src/main/java/io/eventuate/tram/consumer/http/EventuateTramHttpMessageController.java#L19

Current PR does not change this mechanism, however adds mapping between generated id and local id.
Subscription inside proxy service still identified via generated unique identifier, but message is passed using user defined local id (callbackSubscriptionId)

### CODE REVIEW

**Comment by cer** [comment]
> Rest consumer is declared using properties

Is this configuration for the proxy service or for the consuming service?

**Comment by dartartem** [comment]
> Is this configuration for the proxy service or for the consuming service?

consumer, here is example from tests:

properties:
https://github.com/eventuate-tram/eventuate-tram-messaging-http/blob/97130266ef2f44b41e601523f4c98c0c645997d7/eventuate-tram-rest-consumer-spring/src/test/resources/application.properties

controller:
https://github.com/eventuate-tram/eventuate-tram-messaging-http/blob/97130266ef2f44b41e601523f4c98c0c645997d7/eventuate-tram-rest-consumer-spring/src/test/java/io/eventuate/tram/rest/consumer/spring/TestController.java

**Comment by cer** [comment]
> subscription.message.s1.callbackSubscriptionId=SomeSubscription

Why not use the subscriberId in the URL?

**Comment by cer** [comment]
Regarding property names:

```
subscription.message.s1.subscriberId=SomeSubscriber
subscription.message.s1.channels=test-channel
subscription.message.s1.callbackSubscriptionId=SomeSubscription
eventuate.rest.consumer.base.url=http://${DOCKER_HOST_IP:localhost}:8081/messages
```

* They should all be prefixed with `eventuate.`
* Not sure why `callbackSubscriptionId` is necessary. It can be `subscriberId`
* Also, not sure why `s1` (third id!!) is required: `eventuate.subscription.message.SomeSubscriber.channels=...`
* `eventuate.rest.consumer.base.url` -> `eventuate.message.consumer.base.url` would be more better (perhaps)

**Comment by dartartem** [comment]
> Why not use the subscriberId in the URL?

Chris, I do not understand why it is useful. Properties is used to send initial subscription info to proxy, including subscriber id.
But subscription is identifies by subscription id. Subscription id is used in url.

In other words, subscriber id is only needed to call subscribe of https://github.com/eventuate-tram/eventuate-tram-core/blob/master/eventuate-tram-consumer-common/src/main/java/io/eventuate/tram/consumer/common/MessageConsumerImplementation.java#L9

But message handler will be associated with unique subscription id, that will be mapped to callbackSubscriptionId,
that is used to associate message handler on proxy side with rest consumer.

**Comment by dartartem** [comment]
>     They should all be prefixed with eventuate.
I will fix

> Not sure why callbackSubscriptionId is necessary. It can be subscriberId

We introduced subscription id on proxy side to identify each call of subscribe (https://github.com/eventuate-tram/eventuate-tram-core/blob/master/eventuate-tram-consumer-common/src/main/java/io/eventuate/tram/consumer/common/MessageConsumerImplementation.java#L9), because it can be done multiple times with the same subscriber id. It is not possible to use generated subscription id directly, because it is generated during subscription, so I used that concept.


> Also, not sure why s1 (third id!!) is required: eventuate.subscription.message.SomeSubscriber.channels=...

It is only for grouping, but ok, I will remove it

> eventuate.rest.consumer.base.url -> eventuate.message.consumer.base.url would be more better (perhaps)

How about eventuate.rest.message.consumer.base.url ?

**Comment by cer** [comment]
> How about eventuate.rest.message.consumer.base.url ?

I don't see the point in specifying `rest`. There are no alternative values. And, this is not a REST API anyway.

**Comment by dartartem** [comment]
> I don't see the point in specifying rest. There are no alternative values. And, this is not a REST API anyway.

ok

**Comment by cer** [comment]
> We introduced subscription id on proxy side to identify each call of subscribe (https://github.com/eventuate-tram/eventuate-tram-core/blob/master/eventuate-tram-consumer-common/src/main/java/io/eventuate/tram/consumer/common/MessageConsumerImplementation.java#L9), because it can be done multiple times with the same subscriber id. It is not possible to use generated subscription id directly, because it is generated during subscription, so I used that concept.

Not sure exactly what you are trying to say.
However, `subscriberId`  (as passed to MessageConsumer.subscribe()) is a key Eventuate concept and appropriate to have in the URL path.
`SubscriptionId` seems like an internal detail that does not belong in the path.

**Comment by dartartem** [comment]
> Not sure exactly what you are trying to say.
> However, subscriberId (as passed to MessageConsumer.subscribe()) is a key Eventuate concept and appropriate to have in the URL path.
> SubscriptionId seems like an internal detail that does not belong in the path.

Chris,

could you please clarify 2 points:

1) I just need to add subscriber id to path and do not remove it from properties, right?

2) should I remove `callbackSubscriptionId` from properties?

**Comment by cer** [comment]
> could you please clarify 2 points:
> 
> 1. I just need to add subscriber id to path and do not remove it from properties, right?
> 2. should I remove `callbackSubscriptionId` from properties?

* Property `subscription.message.SomeSubscriber.channels=x,y,z`
* `@PostMapping(path = "/messages/SomeSubscription")`

**Comment by dartartem** [comment]
>     Property subscription.message.SomeSubscriber.channels=x,y,z
>     @PostMapping(path = "/messages/SomeSubscription")

I am sorry, but I do not understand what do you mean.

**Comment by cer** [comment]
One more suggestion: The proxy POSTs to `/messages/<subscriberId>/<channel>` - there could be one or many @Controller methods.

**Comment by cer** [comment]
> > ```
> > Property subscription.message.SomeSubscriber.channels=x,y,z
> > @PostMapping(path = "/messages/SomeSubscription")
> > ```
> 
> I am sorry, but I do not understand what do you mean.

This is your example: https://github.com/eventuate-tram/eventuate-tram-messaging-http/pull/8#issuecomment-791530833

I am suggesting that it become:

```
eventuatesubscription.message.SomeSubscriber.channels=x,y,z
eventuate.message.consumer.base.url=http://${DOCKER_HOST_IP:localhost}:8081/messages
```

And 

```
@RestController
public class TestController {
  @PostMapping(path = "/messages/SomeSubscriber/*")
  public void handleMessages(@RequestBody HttpMessage httpMessage) {
    ...
  }
}
```

**Comment by dartartem** [comment]
> This is your example: #8 (comment)
> 
> I am suggesting that it become:
> 
> eventuatesubscription.message.SomeSubscriber.channels=x,y,z
> eventuate.message.consumer.base.url=http://${DOCKER_HOST_IP:localhost}:8081/messages
> 
> And
> 
> @RestController
> public class TestController {
>   @PostMapping(path = "/messages/SomeSubscription/*")
>   public void handleMessages(@RequestBody HttpMessage httpMessage) {
>     ...
>   }
> }

Probably you meant

`@PostMapping(path = "/messages/SomeSubscriber/*")`

not

`@PostMapping(path = "/messages/SomeSubscription/*")`

?

**Comment by dartartem** [comment]
Added support for events.

Controller:

https://github.com/eventuate-tram/eventuate-tram-messaging-http/blob/0f08782308c7faf9f92a2084f9e58ceb928b865e/eventuate-tram-rest-consumer-spring/src/test/java/io/eventuate/tram/rest/consumer/spring/TestController.java#L29-L32

properties:
https://github.com/eventuate-tram/eventuate-tram-messaging-http/blob/0f08782308c7faf9f92a2084f9e58ceb928b865e/eventuate-tram-rest-consumer-spring/src/test/resources/application.properties#L10-L11

**For now controller uses event as is (without envelope), should I add it?**

Also made individual urls for all subscriptions:

https://github.com/eventuate-tram/eventuate-tram-messaging-http/blob/0f08782308c7faf9f92a2084f9e58ceb928b865e/eventuate-tram-rest-consumer-spring/src/test/resources/application.properties#L1-L11

**Comment by cer** [comment]
> **For now controller uses event as is (without envelope), should I add it?**

What problem does adding an envelop solve? And what are the alternative solutions?

**Comment by cer** [comment]
> eventuate.subscription.event.4.aggregates=TestAggregate 
>  eventuate.subscription.event.4.url=http://${DOCKER_HOST_IP:localhost}:8081/test-event

This doesn't seem quite right. An aggregate can have multiple event types - each of which needs a different URL path.

I think a better scheme for URL paths is:

* message - /subscriberId/channel
* event - /subscriberId/aggregate/aggregateId/eventType/eventId

That way you can write

```
 @PostMapping(path = "/mySubscriberId/Order/{orderId}/orderCreated/{eventId}") 
public void handleOrderCreatedEvent(@PathVariable String orderId, @RequestBody TestEvent testEvent) { 
```

**Comment by dartartem** [comment]
> What problem does adding an envelop solve? And what are the alternative solutions?
It contains additional information:

```
  String getAggregateId();
  Message getMessage();
  String getAggregateType();
  String getEventId();
```

I think a good solution would be to add property that allows to use envelopes instead of pure events.

I think it is example of good a api:

> a good API design principle is to make common cases as easy as possible and while preserving the ability to uncommon things

**Comment by dartartem** [comment]
> This doesn't seem quite right. An aggregate can have multiple event types - each of which needs a different URL path.
> ...

I will fix

File: eventuate-tram-messaging-proxy-service/src/test/resources/application.properties [hunk 1]
```diff
@@ -1 +1,28 @@
 eventuate.http.proxy.zookeeper.connection.string=${DOCKER_HOST_IP:localhost}:2181
+eventuatelocal.kafka.bootstrap.servers=${DOCKER_HOST_IP:localhost}:9092
+
+eventuate.subscription.message.1.channels=orders,order-history
```
Comment (cer) [inline eventuate-tram-messaging-proxy-service/src/test/resources/application.properties]: `subscriberId` should be a string, not a number.

File: eventuate-tram-messaging-proxy-service/src/test/resources/application.properties [hunk 2]
```diff
@@ -1 +1,28 @@
...
+eventuate.subscription.message.1.url=http://${DOCKER_HOST_IP:localhost}:8081/messages/1
+
+eventuate.subscription.message.2.channels=customers,customer-history
+eventuate.subscription.message.2.url=http://${DOCKER_HOST_IP:localhost}:8081/messages/2
```
Comment (cer) [inline eventuate-tram-messaging-proxy-service/src/test/resources/application.properties]: `eventuate.subscription.message.2.url` should not include the `subscriberId`, i.e. '2'.
The proxy should automatically include that in the path.

File: eventuate-tram-messaging-proxy-service/src/test/resources/application.properties (hunk 1 above)
Comment (dartartem) [reply eventuate-tram-messaging-proxy-service/src/test/resources/application.properties]: internally it is string I just used number in tests. I will change to some text values.

File: eventuate-tram-messaging-proxy-service/src/test/resources/application.properties (hunk 2 above)
Comment (dartartem) [reply eventuate-tram-messaging-proxy-service/src/test/resources/application.properties]: ok, I will fix

**Pull request merged**